### PR TITLE
Insufficient funds handling

### LINKED ISSUE #2222: Explorer: checkout on optimism allows checking out with DAI even if the users's balance is zero
**Overview**
Details, details, details

**Steps To Reproduce**

**Observed  behavior**
A clear and concise description of what the observed behavior of the system.

**Expected behavior**
A clear and concise description of what you expected the system to do.

**Screenshots**

**Additional context**
Add any other context about the problem here that will assist in triaging the bug. I.e. Browser information, unique identifiers, etc.

### PR DESCRIPTION
**Pull request by vacekj**
This PR addresses several things:
- re-enables the insufficient funds dialog per-chain
- omits the chains with insufficient funds from checkout, but allows checking out other chains

## PR checklist

- Tests postponed until refactor of this component into smaller parts
- [x] Let the user donate to chains they have funds on - disable the insufficient chains in the chain confirmation dialog

##### Refers/Fixes

fixes #2222 #2226 #2403

### CODE REVIEW

**Comment by michellema1208** [comment]
@vacekj love the new error handling here - great work!! Only thing I would add is that we should disable the "Submit your donation!" button when you have insufficient funds 

Screen Shot 2023-12-06 at 3 02 57 PM

**Comment by melissa-neira** [comment]
+1 to @michellema1208's comments. if there is only one network in the cart and there's insufficient funds, then the submit donation button should be disabled. but, if there are other networks in the cart with sufficient funds, then the user should be able to donate, just not on the network with insufficient funds

File: packages/grant-explorer/src/features/round/ViewCartPage/SummaryContainer.tsx [hunk 1]
```diff
@@ -43,6 +47,71 @@ export function SummaryContainer() {
...
+            chainId,
+          });
+          newMap.set(chainId, value);
+        })
+      );
+      setTokenBalancesPerChain(newMap);
+    };
+    runner();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [address, chainIds, getVotingTokenForChain]);
```
Comment (carlbarrdahl) [inline packages/grant-explorer/src/features/round/ViewCartPage/SummaryContainer.tsx:80]: The pattern of combining `useState` and `useEffect` can often be replaced with a `useSWR` hook. That way we get caching (memo), error and loading states.

**Review by meglister — APPROVED** [review]

File: packages/grant-explorer/src/features/round/ViewCartPage/SummaryContainer.tsx (hunk 1 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/round/ViewCartPage/SummaryContainer.tsx:80]: will do in next iteration, don't wanna change too much until we have tests for this page, but very much agree on that.

File: packages/grant-explorer/src/features/round/ViewCartPage/Summary.tsx [hunk 2]
```diff
@@ -22,36 +24,61 @@ export function Summary({
...
+
+  const { address } = useAccount();
+
+  const { data: balance } = useBalance({
+    address,
+    token:
+      selectedPayoutToken.address === zeroAddress
+        ? undefined
+        : selectedPayoutToken.address,
+    chainId,
```
[the remark this answers is not in the document]
Comment (bard) [reply packages/grant-explorer/src/features/round/ViewCartPage/Summary.tsx:36]: In that case, please avoid having the reader process the concept of coercing a string into a boolean and use the explicit, unambiguous check: `enabled: address !== undefined`.

File: packages/grant-explorer/src/features/round/ViewCartPage/ChainConfirmationModalBody.tsx [hunk 3]
```diff
@@ -42,20 +42,24 @@ export function ChainConfirmationModalBody({
...
-            totalDonation={totalDonationsPerChain[Number(chainId)]}
-            checked={chainIdsBeingCheckedOut.includes(Number(chainId))}
-            onChange={(checked) =>
-              handleChainCheckboxChange(Number(chainId), checked)
-            }
-            isLastItem={index === Object.keys(projectsByChain).length - 1}
-          />
-        ))}
+        {Object.keys(projectsByChain)
+          .filter((chainId) =>
```
Comment (bard) [inline packages/grant-explorer/src/features/round/ViewCartPage/ChainConfirmationModalBody.tsx:46]: Would the following make the conversions and coercions unnecessary?

```ts
import { parseChainId } from 'common/src/chains';
// ...
  {Object.keys(projectsByChain)
    .map(parseChainId)
    .filter(...
```

File: packages/grant-explorer/src/features/round/ViewCartPage/SummaryContainer.tsx [hunk 4]
```diff
@@ -43,6 +47,71 @@ export function SummaryContainer() {
     [projects]
   );
 
+  const chainIds = useMemo(
+    () => Object.keys(projectsByChain).map(Number),
+    [projectsByChain]
+  );
```
Comment (bard) [inline packages/grant-explorer/src/features/round/ViewCartPage/SummaryContainer.tsx:53]: Unless the processed data is big, the performance increase in using `useMemo` doesn't outweigh the decrease in readability.

**Review by melissa-neira — APPROVED** [review]
lgtm!

File: packages/grant-explorer/src/features/round/ViewCartPage/Summary.tsx [hunk 5]
```diff
@@ -22,36 +24,61 @@ export function Summary({
...
+  });
+
+  const insufficientFunds = balance ? totalDonation > balance.value : false;
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/features/round/ViewCartPage/Summary.tsx:39]: Added a todo for next iteration.

File: packages/grant-explorer/src/features/round/ViewCartPage/ChainConfirmationModalBody.tsx (hunk 3 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/round/ViewCartPage/ChainConfirmationModalBody.tsx:46]: fixed.

**Pull request merged**